pbft/pbft.cpp

```diff
@@ -469,4 +491,130 @@ pbft::handle_failure()
+
+    pbft_msg cp_msg;
+    cp_msg.set_type(PBFT_MSG_CHECKPOINT);
+    cp_msg.set_view(this->view);
+    cp_msg.set_sequence(sequence);
+    cp_msg.set_sender(this->uuid);
+    cp_msg.set_state_hash(cp->second);
+
+    this->broadcast(this->wrap_message(cp_msg));
+
+    this->maybe_stabilize_checkpoint(*cp);
+}
+
+void
+pbft::handle_checkpoint(const pbft_msg& msg)
+{
+    if (msg.sequence() <= stable_checkpoint.first)
+    {
+        LOG(debug) << boost::format("Ignoring checkpoint message for seq %1% because I already have a stable checkpoint at seq %2%")
+                   % msg.sequence()
+                   % stable_checkpoint.first;
+        return;
+    }
+
+    LOG(info) << boost::format("Recieved checkpoint message for seq %1% from %2%")
+              % msg.sequence()
+              % msg.sender();
+
+    checkpoint_t cp(msg.sequence(), msg.state_hash());
+
+    this->unstable_checkpoint_proofs[cp][msg.sender()] = msg.SerializeAsString();
+    this->maybe_stabilize_checkpoint(cp);
+}
+
+bzn::checkpoint_t
+pbft::latest_stable_checkpoint() const
+{
+    return this->stable_checkpoint;
+}
+
+bzn::checkpoint_t
+pbft::latest_checkpoint() const
+{
+    return this->local_unstable_checkpoints.empty() ? this->stable_checkpoint : *(this->local_unstable_checkpoints.rbegin());
+}
+
+size_t
+pbft::unstable_checkpoints_count() const
+{
+    return this->local_unstable_checkpoints.size();
+}
+
+void
+pbft::maybe_stabilize_checkpoint(const checkpoint_t& cp)
+{
+    if (this->local_unstable_checkpoints.count(cp) == 0 || this->unstable_checkpoint_proofs[cp].size() < this->quorum_size())
+    {
+        return;
+    }
+
+    this->stable_checkpoint = cp;
+    this->stable_checkpoint_proof = this->unstable_checkpoint_proofs[cp];
+
+    LOG(info) << boost::format("Checkpoint %1% at seq %2% is now stable; clearing old data")
+              % cp.second
+              % cp.first;
+
+    // Remove obsolete local checkpoints
+    auto local_start = this->local_unstable_checkpoints.begin();
+    // Iterator to the first unstable checkpoint that's newer than this one. This logic assumes that CHECKPOINT_INTERVAL
+    // is >= 2, otherwise we would have do do something awkward here
+    auto local_end = this->local_unstable_checkpoints.upper_bound(checkpoint_t(cp.first+1, ""));
+    size_t local_removed = std::distance(local_start, local_end);
+    this->local_unstable_checkpoints.erase(local_start, local_end);
+    LOG(debug) << boost::format("Cleared %1% unstable local checkpoints") % local_removed;
+
+
+    // Remove obsolete checkpoint proof collections
+    auto start = this->unstable_checkpoint_proofs.begin();
+    auto end = this->unstable_checkpoint_proofs.upper_bound(checkpoint_t(cp.first+1, ""));
+    size_t to_remove = std::distance(start, end);
+    this->unstable_checkpoint_proofs.erase(start, end);
+    LOG(debug) << boost::format("Cleared %1% unstable checkpoint proof sets") % to_remove;
+
+
```

Comment:
There's a few things going on in here. Is it worth breaking out to separate functions?